design/toolchain.md

```diff
@@ -0,0 +1,115 @@
+
+```
++-----+                +---------+             +----------+             +------------+
+| App |-- synthesis -->| Staging |-- bundle -->| Assembly |-- deploy -->| Deployment |
++-----+                +---------+             +----------+             +------------+
+```
+
+### cdk-synth
+
+We start from a ready-to-run (__compiled__) CDK app.
+
+Some languages require compilation (such as Java, Go, TypeScript) and in others (JavaScript, Ruby, Python), there is no need to compile. Compiling code is __not in scope__ for the CDK toolchain in order to allow developers to use their favorite IDEs and build toolchain idiomatically.
+
+The first component in the toolchain is called `cdk-synth`. It's purpose is to execute the CDK app so it can synthesize the CDK tree (through the internal app phases which are "init", "prepare", "validate", "synthesize"). Eventually the CDK tree can synthesize both final artifacts such as CloudFormation templates and intermediate artifacts such as unpackaged assets.
+
+The protocol between the CDK app `cdk-synth`:
+
+- Context is passed into the app via a JSON map serialized into an environment variable called `CONTEXT_ENV`.
+- The `CDK_OUTDIR` points to a directory into which artifacts should be emitted by the app.
+
+The app will emit into `CDK_OUTDIR` a directory structure as follows:
+
+```
+.
+├── missing.json
+├── bundle.json
+├── assembly
+│   ├── manifest.json
+│   └── stack1.template.json
+└── staging
+    └── intermediate-asset
+        └── file1.txt
+```
+
+If the file `missing.json` exists in the output, it means that the synthesis output cannot be used and relies on missing context information, which must be populated in the context map. `cdk-synth` is responsible to resolve this context information through the environmental context provider system (for example, query the account for which availability zones are available in a specific region, read SSM parameters, etc). Then, `cdk-synth` must re-invoke the app with this context. Bear in mind that normally context information is cached in `cdk.context.json`, so subsequent executions will not require resolution.
+
+`cdk-synth` may repeat this process up to 5 times (theoretically there could be missing context that the app will only know that it needs after it used some information from a resolved context from previous iteration).
+
+Apart from reading `missing.json`, `cdk-synth` does not care about the structure of the output directory.
+
+The output directory is expected to be *self-contained* in a sense that it does not rely on any files that may have only been available during the CDK app execution (e.g. files that are embedded as resources in library dependencies and cannot be accessed from outside the runtime (see [#1716](https://github.com/awslabs/aws-cdk/issues/1716)). It should be possible to pick up the output directory produced by "cdk-synth" and execute "cdk-bundle" (the next step) on a separate machine, possible with a different environment.
```

Comment:
`s/possible/possibly/`

Reply:
Also, does the staging directory, or does it not, refer to user files outside the staging directory?

Does it contain `build instructions` only, or `build instructions+build source`?

Reply:
It contains build instructions + source. I'll make it clear